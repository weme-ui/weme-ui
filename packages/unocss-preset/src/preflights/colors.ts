import type { Preflight, WemeColorScales, WemePresetOptions } from '../types'
import { minifyCss, trackedColors } from '../utils'

export function preflightColors(options: WemePresetOptions): Preflight {
  const colors = options.colors

  return {
    layer: 'theme',
    getCSS() {
      if (trackedColors.size === 0)
        return ''

      const light: string[] = []
      const dark: string[] = []

      Array.from(trackedColors).forEach((key) => {
        const [name, scale] = key.split('-')

        if (Object.keys(colors).includes(name)) {
          const color = colors[name]

          light.push(`--${key}: ${color.light[scale as any as keyof WemeColorScales]};`)
          dark.push(`--${key}: ${color.dark[scale as any as keyof WemeColorScales]};`)
        }
      })

      return minifyCss(`:root {
${light.join('\n')}
}

:root:where(.dark) {
${dark.join('\n')}
}
`)
    },
  }
}
