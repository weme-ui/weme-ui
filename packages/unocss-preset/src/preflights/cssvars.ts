import type { Preflight, WemePresetOptions } from '../types'
import { createCssVars, minifyCss } from '../utils'

export function preflightCssVars(prefix: string, cssVars: WemePresetOptions['cssVars']): Preflight {
  return {
    getCSS() {
      return minifyCss(`
        :root {
          ${createCssVars(prefix, cssVars)}
        }
      `)
    },
    layer: 'base',
  }
}
