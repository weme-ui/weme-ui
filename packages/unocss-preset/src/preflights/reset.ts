import type { Preflight } from '../types'
import { minifyCss } from '../utils'

const resetCSS = `
html {
  font-size: 16px;
  scroll-behavior: smooth;
  overflow-wrap: break-word;
  text-size-adjust: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
`

export function preflightReset(): Preflight {
  return {
    getCSS() {
      return minifyCss(resetCSS)
    },
    layer: 'base',
  }
}
