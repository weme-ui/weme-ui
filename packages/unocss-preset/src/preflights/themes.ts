import type { ColorScalesArray } from '@weme-ui/colors'
import type { Preflight, WemeTheme } from '../types'
import { createCssVars, minifyCss } from '../utils'

export function preflightThemes(prefix: string, themes: WemeTheme[]): Preflight[] {
  return Object.entries(themes)
    .map(([_, theme]) => {
      return {
        getCSS() {
          return minifyCss(`
            [data-theme=${theme.id}] {
              color-scheme: ${theme.appearance};
              --${prefix}radius: ${theme.radius};
              ${createThemeColor(prefix, theme.colors)}
              ${createCssVars(prefix, theme.tokens as unknown as Record<string, string>)}
            }
          `)
        },
        layer: 'base',
      }
    })
}

function createThemeColor(prefix: string, colors: WemeTheme['colors']): string {
  return Object.entries(colors)
    .map(([name, scales]: [string, ColorScalesArray]) => {
      // support custom color scales
      if (['accent', 'neutral'].includes(name)) {
        return [
          // default
          `--${prefix}${name}: var(--${prefix}${name}-9);`,
          // scales
          ...scales.map(
            (value, i) => `--${prefix}${name}-${i + 1}: var(--custom-${name}-${i + 1}, ${value});`,
          ),
        ].join('')
      }

      return [
        // default
        `--${prefix}${name}: var(--${prefix}${name}-9);`,
        // scales
        ...scales.map(
          (value, i) => `--${prefix}${name}-${i + 1}: ${value};`,
        ),
      ]
    })
    .join('')
}
