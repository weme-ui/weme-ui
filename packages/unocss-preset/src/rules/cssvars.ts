/* eslint-disable regexp/no-empty-capturing-group */
/* eslint-disable regexp/no-empty-group */
import type { Rule, WemePresetOptions } from '../types'

export function cssVarRules(options: WemePresetOptions): Rule[] {
  return [
    // Background
    createCssVarRule(
      /^bg-(.+)$/,
      'background',
      options,
    ),

    // Foreground
    createCssVarRule(
      /^(?:c|text)-(.+)$/,
      'color',
      options,
    ),

    // Width
    createSizeCssVarRule(
      /^w-(.+)$/,
      'width',
      options,
    ),

    // Height
    createSizeCssVarRule(
      /^h-(.+)$/,
      'height',
      options,
    ),

    // Size
    createSizeCssVarRule(
      /^size-(.+)$/,
      'size',
      options,
    ),

    // Padding
    createPaddingCssVarRule(
      /^pa?()-?(.+)$/,
      'padding',
      options,
    ),
    createPaddingCssVarRule(
      /^p-?([rltbsexy])(?:-?(.+))?$/,
      'padding',
      options,
    ),

    // Margin
    createPaddingCssVarRule(
      /^ma?()-?(.+)$/,
      'margin',
      options,
    ),
    createPaddingCssVarRule(
      /^m-?([rltbsexy])(?:-?(.+))?$/,
      'margin',
      options,
    ),
  ]
}

function createCssVarRule(matcher: RegExp, property: string, options: WemePresetOptions): Rule {
  const { prefix, cssVars } = options

  const map: Record<string, Record<string, string>> = {
    token: {
      color: 'foreground',
    },
    suffix: {
      color: 'fg',
      background: 'bg',
    },
  }

  return [
    matcher,
    ([, c]) => {
      const suffix = map.suffix[property] ?? property

      if (Object.keys(cssVars).includes(`${c}-${map.token[property] ?? property}`)) {
        return {
          [`${property}`]: `var(--${prefix}${c}-${suffix})`,
        }
      }
    },
  ]
}

function createSizeCssVarRule(matcher: RegExp, property: string, options: WemePresetOptions): Rule {
  const { prefix, cssVars } = options

  if (property === 'size') {
    return [
      matcher,
      ([, c]) => {
        if (Object.keys(cssVars).includes(`${c}-width`) && Object.keys(cssVars).includes(`${c}-height`)) {
          return {
            width: `var(--${prefix}${c}-width)`,
            height: `var(--${prefix}${c}-height)`,
          }
        }

        if (Object.keys(cssVars).includes(`${c}-width`)) {
          return {
            width: `var(--${prefix}${c}-width)`,
            height: `var(--${prefix}${c}-width)`,
          }
        }

        if (Object.keys(cssVars).includes(`${c}-height`)) {
          return {
            width: `var(--${prefix}${c}-height)`,
            height: `var(--${prefix}${c}-height)`,
          }
        }
      },
    ]
  }

  return createCssVarRule(matcher, property, options)
}

function createPaddingCssVarRule(matcher: RegExp, property: string, options: WemePresetOptions): Rule {
  const { prefix, cssVars } = options

  const map: Record<string, Record<string, string>> = {
    direction: {
      l: 'left',
      r: 'right',
      t: 'top',
      b: 'bottom',
      s: 'left',
      e: 'right',
      x: 'inline',
      y: 'block',
    },
  }

  if (property === 'padding' || property === 'margin') {
    return [
      matcher,
      ([, side, c]) => {
        const matched = Object.keys(cssVars)
          .filter(k => k.startsWith(`${c}-${property}`))

        if (matched.length > 0) {
          const value = matched.find(k => k === `${c}-${property}-${map.direction[side]}`)
            ? `var(--${prefix}${c}-${property}-${map.direction[side]})`
            : `var(--${prefix}${c}-${property})`

          if (side === 'l' || side === 'r' || side === 't' || side === 'b') {
            return {
              [`${property}-${map.direction[side]}`]: value,
            }
          }
          else if (side === 's') {
            return {
              [`${property}-inline-start`]: value,
            }
          }
          else if (side === 'e') {
            return {
              [`${property}-inline-end`]: value,
            }
          }
          else if (side === 'x') {
            return {
              [`${property}-inline`]: value,
            }
          }
          else if (side === 'y') {
            return {
              [`${property}-block`]: value,
            }
          }

          return {
            [`${property}`]: `var(--${prefix}${c}-${property})`,
          }
        }
      },
    ]
  }

  return createCssVarRule(matcher, property, options)
}
