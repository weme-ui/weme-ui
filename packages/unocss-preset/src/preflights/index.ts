import type { Preflight, WemePresetOptions } from '../types'
import { preflightColors } from './colors'
import { preflightCssVars } from './cssvars'
import { preflightProperties } from './properties'
import { preflightReset } from './reset'
import { preflightThemes } from './themes'

export function preflights(options: WemePresetOptions): Preflight[] {
  return [
    ...preflightThemes(options),
    preflightProperties(),
    preflightReset(options.variablePrefix),
    preflightColors(options),
    preflightCssVars(options),
  ].filter(Boolean) as Preflight[]
}
