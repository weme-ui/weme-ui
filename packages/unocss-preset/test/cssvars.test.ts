import { describe, expect, it } from 'vitest'
import { preflightCssVars } from '../src/preflights/cssvars'

describe('css variables', () => {
  it('should preflight css variables', () => {
    const cssVars = preflightCssVars(
      'test-',
      {
        background: {
          default: '#fff',
          dimmed: '#eee',
        },
        panel: {
          default: '#eee',
          background: '#eee',
          foreground: '#333',
        },
      },
    )

    expect(cssVars.getCSS({} as any)).toMatchSnapshot()
  })
})
