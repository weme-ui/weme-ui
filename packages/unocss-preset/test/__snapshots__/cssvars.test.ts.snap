// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css variables > should preflight css variables 1`] = `":root { --test-bg: #fff;--test-bg-dimmed: #eee;--test-panel: #eee;--test-panel-bg: #eee;--test-panel-fg: #333; }"`;
